examples/rl/pendulum/ex_tensorboard.py

```diff
@@ -0,0 +1,77 @@
+import functools
+from datetime import datetime
+
+import jax
+from brax import envs
+from brax.training.agents.ppo import networks as ppo_networks
+from brax.training.agents.ppo import train as ppo
+from tensorboardX import SummaryWriter
+
+from ambersim.rl.pendulum.swingup import PendulumSwingupEnv
+
+"""
+A pendulum swingup example that uses a tensorboard callback to log training
+progress in real time.
+
+You can view the logs with
+$ tensorboard --logdir /tmp/ambersim
+"""
+
+if __name__ == "__main__":
+    # Initialize the environment
+    envs.register_environment("pendulum_swingup", PendulumSwingupEnv)
+    env = envs.get_environment("pendulum_swingup")
+
+    # Define the training function
+    network_factory = functools.partial(
+        ppo_networks.make_ppo_networks,
+        policy_hidden_layer_sizes=(64,) * 3,
+    )
+    train_fn = functools.partial(
+        ppo.train,
+        num_timesteps=100_000,
+        num_evals=50,
+        reward_scaling=0.1,
+        episode_length=200,
+        normalize_observations=True,
+        action_repeat=1,
+        unroll_length=10,
+        num_minibatches=32,
+        num_updates_per_batch=8,
+        discounting=0.97,
+        learning_rate=3e-4,
+        entropy_cost=0,
+        num_envs=1024,
+        batch_size=512,
+        network_factory=network_factory,
+        seed=0,
+    )
+
+    # Set up a tensorboard logger
+    log_dir = "/tmp/ambersim/pendulum"
+    print(f"Setting up Tensorboard logging in {log_dir}")
+    writer = SummaryWriter(log_dir)
+
+    # Define a callback to log progress
+    times = [datetime.now()]
+
+    def progress(num_steps, metrics):
+        """Logs progress during RL."""
+        print(f"  Steps: {num_steps}, Reward: {metrics['eval/episode_reward']}")
+        times.append(datetime.now())
+
+        # Write all metrics to tensorboard
+        for key, val in metrics.items():
+            if isinstance(val, jax.Array):
+                val = float(val)  # we need floats for logging
+            writer.add_scalar(key, val, num_steps)
```

Comment:
Is it possible to write a util the autogenerates a progress callback with a fixed API that also auto-logs to tensorboard? Since `metrics` is totally customizable on the user side, we can maybe just iterate through all of them and publish.

I'm envisioning a new `utils` file called maybe `logging_utils.py`? Maybe `vis_utils.py`? Open to suggestions/discussion.

If you don't think it's a good idea or you want to punt this util, feel free to open an issue instead and merge.

Reply:
I do think that's a good idea: opened #37